thetis/exner_eq.py

```diff
@@ -0,0 +1,141 @@
+    There are no boundary conditions for the Exner equation.
+    """
+    def __init__(self, function_space, depth, sediment_model, depth_integrated_sediment=False):
+        """
+        :arg function_space: :class:`FunctionSpace` where the solution belongs
+        :arg depth: :class: `DepthExpression` containing depth info
+        :arg sediment_model: :class: `SedimentModel` containing sediment info
+        :kwarg bool depth_integrated_sediment: whether the sediment field is depth-integrated
+        """
+        super(ExnerTerm, self).__init__(function_space)
+        self.n = FacetNormal(self.mesh)
+        self.depth = depth
+
+        self.sediment_model = sediment_model
+
+        # define measures with a reasonable quadrature degree
+        p = self.function_space.ufl_element().degree()
+        self.quad_degree = 2*p + 1
+        self.dx = dx(degree=self.quad_degree)
+        self.dS = dS(degree=self.quad_degree)
+        self.ds = ds(degree=self.quad_degree)
+        self.depth_integrated_sediment = depth_integrated_sediment
+
+
+class ExnerSourceTerm(ExnerTerm):
+    r"""
+    Source term accounting for suspended sediment transport
+
+    The weak form reads
+
+    .. math::
+        F_s = \int_\Omega (\sigma - sediment * \phi) * depth \psi dx
+
+    where :math:`\sigma` is a user defined source scalar field :class:`Function`
+    and :math:`\phi` is a user defined source scalar field :class:`Function`.
+    """
+    def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=None):
+
+        f = 0
+        sediment = fields.get('sediment')
+        morfac = fields.get('morfac')
+        porosity = fields.get('porosity')
+
+        fac = Constant(morfac/(1.0-porosity))
+        H = self.depth.get_total_depth(fields_old['elev_2d'])
+
+        erosion = self.sediment_model.get_erosion_term()
+        deposition = self.sediment_model.get_deposition_coefficient() * sediment
+        if self.depth_integrated_sediment:
+            deposition = deposition/H
+        f = self.test*fac*(erosion - deposition)*self.dx
+
+        return f
+
+
+class ExnerBedloadTerm(ExnerTerm):
+    r"""
+    Bedload transport term, \nabla_h \cdot \textbf{qb}
+
+    The weak form is
+
+    .. math::
+        \int_\Omega  \nabla_h \cdot \textbf{qb} \psi  dx
+        = - \int_\Omega (\textbf{qb} \cdot \nabla) \psi dx
+        + \int_\Gamma \psi \textbf{qb} \cdot \textbf{n} dS
+
+    where :math:`\textbf{n}` is the unit normal of the element interfaces.
+    """
+    def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=None):
+        f = 0
+
+        qbx, qby = self.sediment_model.get_bedload_term(solution)
+
+        morfac = fields.get('morfac')
+        porosity = fields.get('porosity')
+
+        fac = Constant(morfac/(1.0-porosity))
+
+        f += -(self.test*((fac*qbx*self.n[0]) + (fac*qby*self.n[1])))*self.ds(1)
+        f += -(self.test*((fac*qbx*self.n[0]) + (fac*qby*self.n[1])))*self.ds(2)
```

Comment:
There's some hard-coded surface ids in there! Do we need boundary conditions for the Exner equation?